File: examples/p3_oop_animals.py
#!/usr/bin/env python3
'''
@author  Michele Tomaiuolo - http://www.ce.unipr.it/people/tomamic
@license This software is free - http://www.gnu.org/licenses/gpl.html
'''

class Animal:
    def say(self):
        raise NotImplementedError("Abstract method")

class Dog(Animal):
    def __init__(self, name):
        self._name = name

    def say(self):
        print("I am", self._name, "Dog. I say: WOOF!")

class Cat(Animal):
    def __init__(self, name):
        self._name = name

    def say(self):
        print("I am", self._name, "Cat. I say: MEOW!")

class Pig(Animal):
    def __init__(self, name):
        self._name = name

    def say(self):
        print("I am", self._name, "Pig. I say: OINK!")


# a list of Animal objects
d = Dog("Danny")
c = Cat("Candy")
p1 = Pig("Peppa")
p2 = Pig("George")

animals = [d, c, p1, p2]

for a in animals:
    a.say()

File: go/p3_oop_animals.go
package main

import . "g2d"

type Animal interface {
    Say()
}

type Dog struct {
    name string
}

type Cat struct {
    name string
}

type Pig struct {
    name string
}

func NewDog(name string) *Dog {
    return &Dog{name}
}

func NewCat(name string) *Cat {
    return &Cat{name}
}

func NewPig(name string) *Pig {
    return &Pig{name}
}

func (d *Dog) Say() {
    Println("I'm " + d.name + " Dog. I say: WOOF!")
}

func (c *Cat) Say() {
    Println("I'm " + c.name + " Cat. I say: MEOW!")
}

func (p *Pig) Say() {
    Println("I'm " + p.name + " Pig. I say: OINK!")
}

func main() {
    d := NewDog("Danny")
    c := NewCat("Candy")
    p1 := NewPig("Peppa")
    p2 := NewPig("George")
    animals := []Animal{d, c, p1, p2}

    for _, a := range animals {
        a.Say()
    }
}
